[WIP] Support running Postgres in Docker for tests

akheron opened the pull request:
Here's a Dockerfile for running tests against a Postgres that's run in Docker.

To make all tests pass, the Postgres' unix domain socket must be made available in the host via a bind mount volume.

Run postgres:
```sh
cd ci
docker build -t postgres-test .
mkdir socket
chmod 777 socket
docker run --rm -e POSTGRES_PASSWORD=dummy -p 5432:5432 -v $(pwd)/socket:/socket postgres-test:latest
```

Run tests:
```
PGHOST=localhost PGUSER=postgres PGSOCKET=ci/socket npm run test
```

I can try to add a full CI setup if there's interest.

porsager commented:
Hey :) This is really awesome. I've been wanting to do this for a long time, much appreciated, and there's definitely interest! ;)

karlhorky commented:
> there's definitely interest in a full CI setup too

> I suppose it'd be fairly easy to also test different Postgres versions once that is done?

The PostgreSQL Docker on GitHub actions is not too tough to set up (and works with the current version):

https://twitter.com/getupleveled/status/1323318825526435842

This uses https://github.com/Harmon758/postgresql-action in the background, which allows specifying the version number.

akheron commented:
Rebased and added initial GitHub aAtions.

There are 2 problems with the GitHub Actions setup:

- The unix domain socket test doesn't pass, because I couldn't find a way of exposing the PostgreSQL socket to the host machine in GitHub Actions. There's no way to create a directory with certain permissions on the host machine before starting a service. If I let Docker create the directory, the permissions don't allow PostgreSQL to create the socket file.
- It currently uses the docker image built from `ci/Dockerfile` from GitHub packages Docker container registry of a temporary repository of mine. I added a GitHub Actions workflow for creating the image, but it cannot be tested before this PR is merged because forks cannot have GitHub Packages repositories and manually triggered GitHub Actions workflows don't show up before merging it seems.

porsager commented:
So cool @akheron ! Sorry I haven't had time to chime in.

I think it'd be great if we could contain the necessary bootstrapping required in https://github.com/porsager/postgres/blob/master/tests/bootstrap.js so it can be shared between docker and local test runs. 

I'll try to see if I can figure out something wrt. the local socket.

> because forks cannot have GitHub Packages repositories

would it help if we merged to another branch instead of master first?

porsager commented:
I'm not going to jump into this, so I'll close it for now and if anyone feels like trying we can reopen.

karlhorky commented:
The `ubuntu-latest` virtual environment on GitHub Actions actually has PostgreSQL preinstalled, so if all you want to do is set up CI with Postgres.js in GitHub Actions, it's as easy as:

```yml
name: Playwright tests
on: push

jobs:
  build:
    name: Playwright tests
    runs-on: ubuntu-latest
    env:
      PGHOST: localhost
      PGDATABASE: next_js_example_jan_2022
      PGUSERNAME: next_js_example_jan_2022
      PGPASSWORD: next_js_example_jan_2022
    steps:
      - name: Start preinstalled PostgreSQL on Ubuntu
        run: |
          sudo systemctl start postgresql.service
          pg_isready
      - name: Create database user
        run: |
          sudo -u postgres psql --command="CREATE USER next_js_example_jan_2022 PASSWORD 'next_js_example_jan_2022'" --command="\du"
      - name: Create database and allow user
        run: |
          sudo -u postgres createdb --owner=next_js_example_jan_2022 next_js_example_jan_2022
```

Source: https://github.com/upleveled/next-js-example-jan-2022/blob/d1a94b0a6f2d36cd27a16530e07d4c3ce2ab0c41/.github/workflows/test-puppeteer.yml#L10-L40

porsager commented:
PR welcome 😊 

> On 2 Apr 2022, at 13.00, Karl Horky ***@***.***> wrote:
> 
> ﻿
> The ubuntu-latest virtual environment on GitHub Actions actually has PostgreSQL preinstalled, so if all you want to do is set up CI with Postgres.js in GitHub Actions, it's as easy as:
> 
> name: Playwright tests
> on: push
> 
> jobs:
>   build:
>     name: Playwright tests
>     runs-on: ubuntu-latest
>     env:
>       PGHOST: localhost
>       PGDATABASE: next_js_example_jan_2022
>       PGUSERNAME: next_js_example_jan_2022
>       PGPASSWORD: next_js_example_jan_2022
>     steps:
>       - name: Start preinstalled PostgreSQL on Ubuntu
>         run: |
>           sudo systemctl start postgresql.service
>           pg_isready
>       - name: Create database user
>         run: |
>           sudo -u postgres psql --command="CREATE USER next_js_example_jan_2022 PASSWORD 'next_js_example_jan_2022'" --command="\du"
>       - name: Create database and allow user
>         run: |
>           sudo -u postgres createdb --owner=next_js_example_jan_2022 next_js_example_jan_2022
> Source: https://github.com/upleveled/next-js-example-jan-2022/blob/d1a94b0a6f2d36cd27a16530e07d4c3ce2ab0c41/.github/workflows/test-puppeteer.yml#L10-L40
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you modified the open/close state.

porsager commented:
Did not plan to do this, but nice nerd sniping @karlhorky 😂 https://github.com/porsager/postgres/actions

karlhorky commented:
Wow nice work, the workflow looks great! 😍 Got some cool stuff in there - we may need to see if we can use any of it!

https://github.com/porsager/postgres/blob/master/.github/workflows/test.yml

And that green checkmark - an additional trust signal 🙌

![Screen Shot 2022-04-04 at 12 55 43](https://user-images.githubusercontent.com/1935696/161529878-e6c9ba3a-4ea4-4952-8a18-a76fd3df4218.png)

porsager commented:
Thanks a lot ! Haven't used Github Actions before, and it seems there are still som race conditions I need to iron out. 

Funny.. I don't see a checkmark at that place as in your screenshot?

karlhorky commented:
Ah sorry, this is because of [the Refined GitHub Browser Extension](https://github.com/refined-github/refined-github) - lots of cool stuff there!
